server/mscalendar/daily_summary.go

```diff
@@ -120,86 +131,94 @@ func (m *mscalendar) SetDailySummaryEnabled(user *User, enable bool) (*store.Dai
-	)
+func (m *mscalendar) ProcessAllDailySummary(now time.Time) error {
+	isAdmin, err := m.IsAuthorizedAdmin(m.actingUser.MattermostUserID)
 	if err != nil {
 		return err
 	}
+	if !isAdmin {
+		return errors.Errorf("Non-admin user attempting ProcessAllDailySummary %s", m.actingUser.MattermostUserID)
+	}
 
 	dsumIndex, err := m.Store.LoadDailySummaryIndex()
 	if err != nil {
 		return err
 	}
-
-	for _, dsum := range dsumIndex {
-		err = m.processDailySummary(dsum)
-		if err != nil {
-			u := NewUser(dsum.MattermostUserID)
-			m.ExpandMattermostUser(u)
-			m.Logger.Errorf("Error posting daily summary for user %s: %s", u.MattermostUser.Username, err.Error())
-		}
+	if len(dsumIndex) == 0 {
+		return nil
 	}
 
-	err = m.Store.SaveDailySummaryIndex(dsumIndex)
+	err = m.Filter(withSuperuserClient)
 	if err != nil {
 		return err
 	}
 
-	return nil
-}
+	requests := []*remote.ViewCalendarParams{}
+	for _, dsum := range dsumIndex {
+		shouldPost, err := shouldPostDailySummary(dsum, now)
+		if err != nil {
+			m.Logger.Errorf("Error posting daily summary for user %s: %s", dsum.MattermostUserID, err.Error())
+			continue
+		}
+		if !shouldPost {
+			continue
+		}
 
-func (m *mscalendar) GetDailySummary(user *User) (string, error) {
-	return m.getDailySummary(user)
-}
+		start, end := getTodayHoursForTimezone(now, dsum.Timezone)
+		req := &remote.ViewCalendarParams{
+			RemoteID:  dsum.RemoteID,
+			StartTime: start,
+			EndTime:   end,
+		}
+		requests = append(requests, req)
+	}
 
-func (m *mscalendar) processDailySummary(dsum *store.DailySummarySettings) error {
-	shouldPost, err := shouldPostDailySummary(dsum)
+	responses, err := m.client.DoBatchViewCalendarRequests(requests)
 	if err != nil {
 		return err
 	}
-	if !shouldPost {
-		return nil
-	}
 
-	user := NewUser(dsum.MattermostUserID)
-	postStr, err := m.getDailySummary(user)
-	if err != nil {
-		return err
+	mappedPostTimes := map[string]string{}
+	byRemoteID := dsumIndex.ByRemoteID()
+	for _, res := range responses {
+		dsum := byRemoteID[res.RemoteID]
+		if res.Error != nil {
+			m.Logger.Errorf("Error rendering user %s calendar: %s %s", dsum.MattermostUserID, res.Error.Code, res.Error.Message)
+		}
+		postStr, err := views.RenderCalendarView(res.Events, dsum.Timezone)
+		if err != nil {
+			m.Logger.Errorf("Error rendering user %s calendar: %s", dsum.MattermostUserID, err.Error())
+		}
+
+		m.Poster.DM(dsum.MattermostUserID, postStr)
+		mappedPostTimes[dsum.MattermostUserID] = time.Now().Format(time.RFC3339)
 	}
-	m.Poster.DM(user.MattermostUserID, postStr)
+	m.Logger.Infof("Processed daily summary for %d users", len(mappedPostTimes))
 
-	dsum.LastPostTime = timeNowFunc().Format(time.RFC3339)
+	// TODO atomic save index
 	return nil
 }
 
-func (m *mscalendar) getDailySummary(user *User) (string, error) {
-	tz, err := m.GetTimezone(user)
-	if err != nil {
-		return "", err
-	}
+func (m *mscalendar) GetDailySummaryForUser(user *User) (string, error) {
+	return m.getDailySummaryForUser(user)
```

Comment:
Why using a private function? Couldn't we add this functionality directly here?

Reply:
Yes, this function was used elsewhere but now it is not :+1: